Dev perception: Work in progress

nikos-tri opened the pull request:
+@stonier

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6621)
<!-- Reviewable:end -->

RussTedrake commented:
we've talked about having a drake/perception folder (and moving the pose estimator work from systems/estimators/dev over there).  can you please move these to drake/perception/dev?

cc @EricCousineau-TRI @kunimatsu-tri

soonho-tri commented:
1) We need to write `+@user_name` in Reviewable (not in Github). Otherwise Reviewable shows that no gets assigned. 
2) There are tabs in the commits (red `>>` in Reviewable), which should be replaced with spaces. Many people in the team are using [clang-format](https://dx13.co.uk/articles/2015/4/3/Setting-up-git-clang-format.html) to automate formatting tasks.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-KpXkkISnCZJ67Zm_3pS:bjktncq)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> We need to write +@user_name in Reviewable (not in Github). Otherwise Reviewable shows that no gets assigned.

This is not true; it's supposed to work (and usually works).  The problem here was that `@nikos-tri` was not a member of `@RobotLocomotion/tri`.

soonho-tri commented:
> This is not true; it's supposed to work (and usually works). The problem here was that @nikos-triwas not a member of @RobotLocomotion/tri.

I see. I guessed it from the symptom (and you've added @stonier separately). Good to know that it's not the case.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-KpXm0ISp3WPeEWMAQH0:b5i5u2p)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
just checked, and he is a member of @RobotLocomotion/TRI

jwnimmer-tri commented:
Ah, he was on the second page so I missed him.  In which case, I'm not sure what went wrong.  (I almost always add reviewers via github comments, not reviewable comments, so in general it definitely works.)

hongkai-dai commented:
Checkpoint. Do we have a plan to include recurrent neural networks, which has a feedback loop, and is time-dependent?

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/neural_network.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYM0xTtqcXJFvDrw5g:-KpYM0xTtqcXJFvDrw5h:b-cvwx7v) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/neural_network.cc#L28)):*
> ```cpp
> 
> 	DRAKE_THROW_UNLESS( W.size() == b.size() );
> 	DRAKE_THROW_UNLESS( W.size() == layers.size() );
> ```

So `layers` does not include either the input or the output layer? We should document this. I have seen some literature including both the input and the output layers.

---

*[drake/dev_perception/neural_network.h, line 24 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYLBelVy5JAVhEsxBa:-KpYLBelVy5JAVhEsxBb:b4cj161) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/neural_network.h#L24)):*
> ```objc
> 
> 	/// TODO(nikos-tri): Add documentation
> 	explicit NeuralNetwork( std::vector<MatrixX<T>> W, std::vector<VectorX<T>> b,
> ```

Use `const std::vector<MatrixX<T>>&` instead of `std::vector<MatrixX<T>>`? 

---

*[drake/dev_perception/neural_network.h, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYKRBppGj4WDuubL4x:-KpYKRBppGj4WDuubL4y:bhz64mm) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/neural_network.h#L34)):*
> ```objc
> 	int getNumInputs() const;
> 	int getNumOutputs() const;
> 	std::unique_ptr<MatrixX<T>> getWeightMatrix( int index, const systems::Context<T>& context ) const;
> ```

Our coding style prefers starting with Capital letter for function names https://google.github.io/styleguide/cppguide.html#Function_Names

---

*[drake/dev_perception/neural_network.h, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYN9SUMhKWmAvTXDHf:-KpYN9SUMhKWmAvTXDHg:b80j2w0) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/neural_network.h#L58)):*
> ```objc
> 
> 	// Indices for weights and biases, which are stored as params in the Context
> 	std::vector<int> matrix_indices_;
> ```

rename this as `weight_indices_`? `matrix` is too general.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-KpYNRIOdp2GYwDmJApO:b-swl7b5)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
+1 to hongkai's question -> Perhaps the NeuralNetwork here needs renaming to FeedforwardNeuralNetwork.

---

Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-Kp_Jfh7EFS8IiC43XW1:bycay00)*
<!-- Sent from Reviewable.io -->

nikos-tri commented:
Coordinating with Eric about moving the work in estimators. I have created an (empty) base class called NeuralNetwork. My class is now called FeedforwardNeuralNetwork as suggested by Naveen, and if other classes of NNs are built, common code can go in the NeuralNetwork base class.

---

Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYM0xTtqcXJFvDrw5g:-KpwLwP6MEo1EDYRrOhR:b-dslijr) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/feedforward_neural_network.cc#L28)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

So `layers` does not include either the input or the output layer? We should document this. I have seen some literature including both the input and the output layers.
</blockquote></details>

Added a comment. Haven't written the bulk of the documentation yet, though.

---

*[drake/dev_perception/neural_network.h, line 24 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYLBelVy5JAVhEsxBa:-KpwM2DO1CZO7Ov_jiX9:b-5y1hvo) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/neural_network.h#L24)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Use `const std::vector<MatrixX<T>>&` instead of `std::vector<MatrixX<T>>`? 
</blockquote></details>

Good idea, thanks.

---

*[drake/dev_perception/neural_network.h, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpYKRBppGj4WDuubL4x:-KpwM3ufwCY0ZsCcDxPN:b-u4aq1c) ([raw file](https://github.com/robotlocomotion/drake/blob/27e5e2115feb4783edc10db26470e14fa5596a17/drake/dev_perception/neural_network.h#L34)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Our coding style prefers starting with Capital letter for function names https://google.github.io/styleguide/cppguide.html#Function_Names
</blockquote></details>

Followed this suggestion for all functions except the getters, which the style guide says can be get_xyz

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-KpwLb1gbQO3DgX8zjcG:b-n3q545)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/feedforward_neural_network.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx-PS2ktNV0pDH-ig1:-Kpx-PS2ktNV0pDH-ig2:b-r0vh4g) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L23)):*
> ```objc
> 
>   /// TODO(nikos-tri): Add documentation
>   /// The number of layers does not count the input or output layers.
> ```

Sorry I am still not clear on this one. If the number of layers does not count the input or the output layers, namely it only include the hidden layers, then shouldn't `W.size() = layers().size() + 1`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 4 of 7 files at r2.
Review status: 3 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 33 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpxA7Wo0BnzynRhwKKI:-KpxA7Wo0BnzynRhwKKJ:b-99tv5r) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.cc#L33)):*
> ```cpp
>     weight_indices_.push_back(this->DeclareNumericParameter(*(Encode(W[i]))));
> 
>     BasicVector<T> biasVector(b[i]);
> ```

variable names should use lower case letters, connected with underscores, according to the google style guide https://google.github.io/styleguide/cppguide.html#Variable_Names. Same for the other variable names.

---

*[drake/dev_perception/feedforward_neural_network.h, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx7qMo_WGNPJZjVADZ:-Kpx7qMo_WGNPJZjVAD_:bdd1i3d) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L35)):*
> ```objc
>   int get_num_inputs() const;
>   int get_num_outputs() const;
>   std::unique_ptr<MatrixX<T>> get_weight_matrix(
> ```

It is not clear to me why we return `std::unique_ptr` here, I think we can return `MatrixX<T>` here?

---

*[drake/dev_perception/feedforward_neural_network.h, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx8txPloOwh_46BNcM:-Kpx8txPloOwh_46BNcN:bdf620p) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L41)):*
> ```objc
>   std::unique_ptr<systems::BasicVector<T>> Encode(
>       const MatrixX<T>& matrix) const;
>   std::unique_ptr<MatrixX<T>> Decode(
> ```

Document this function? It is unclear what the function does until I read the code. And maybe change the name to `BasicVectorToWeightMatrix` to be more specific? We can decode many things. Same for `Encode` function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 3 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 94 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kq3UwJijQX4C65xuDcq:-Kq3UwJjKv3lzzuokSus:bulepjo) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.cc#L94)):*
> ```cpp
> VectorX<T> FeedforwardNeuralNetwork<T>::relu(VectorX<T> in) const {
>   // TODO(nikos-tri) This function begs to be optimized, somehow -- maybe like
>   // negIndices = any( input < 0 )
> ```

BTW I do not believe Eigen has logical indexing slices like MATLAB / NumPy.
You can certainly get the array of values:
```
using BoolT = decltype(std::declval<T>() > 0); // I think there's a function in Drake for this somewhere...
ArrayX<BoolT> result = in.array() > 0;
```
But I do not believe there is an optimized access using this, so I'm not sure what this will gain you.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

nikos-tri commented:
Review status: 1 of 40 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 33 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KpxA7Wo0BnzynRhwKKI:-KqAHMB3AcQ029FYJ5eB:b-39f7am) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.cc#L33)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

variable names should use lower case letters, connected with underscores, according to the google style guide https://google.github.io/styleguide/cppguide.html#Variable_Names. Same for the other variable names.
</blockquote></details>

I think I got them all this time. Thanks.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 94 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kq3UwJijQX4C65xuDcq:-KqAEdMnnwlImfSuy4Ih:b-tkptn5) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.cc#L94)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW I do not believe Eigen has logical indexing slices like MATLAB / NumPy.
You can certainly get the array of values:
```
using BoolT = decltype(std::declval<T>() > 0); // I think there's a function in Drake for this somewhere...
ArrayX<BoolT> result = in.array() > 0;
```
But I do not believe there is an optimized access using this, so I'm not sure what this will gain you.
</blockquote></details>

I think you're right, there doesn't seem to be a way to do this. Bummer.

---

*[drake/dev_perception/feedforward_neural_network.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx-PS2ktNV0pDH-ig1:-KqADOtzYgKnxurc-VaA:b-g24dz7) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L23)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Sorry I am still not clear on this one. If the number of layers does not count the input or the output layers, namely it only include the hidden layers, then shouldn't `W.size() = layers().size() + 1`?
</blockquote></details>

Not sure if this was the spot where this comment belonged, since I don't see any code that looks similar. Maybe you were looking at an old version of the encoding function, which was adding 2. But, that was not due to the layers, it was due to needing two slots to store the number of rows and columns. But, the new version doesn't do that anymore due to annoyances with casting between AutoDiffXd and int.

---

*[drake/dev_perception/feedforward_neural_network.h, line 35 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx7qMo_WGNPJZjVADZ:-KqAE0RWUv1rox_gEIhZ:be0qmq) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L35)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

It is not clear to me why we return `std::unique_ptr` here, I think we can return `MatrixX<T>` here?
</blockquote></details>

Just to avoid a copy operation.

---

*[drake/dev_perception/feedforward_neural_network.h, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx8txPloOwh_46BNcM:-KqAEJMSnChWVrOTNGbW:b-ordcxk) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L41)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Document this function? It is unclear what the function does until I read the code. And maybe change the name to `BasicVectorToWeightMatrix` to be more specific? We can decode many things. Same for `Encode` function.
</blockquote></details>

Of course, everything will be documented eventually. This is a "dev" feature, still experimental. I have changed the function name to be clearer as you suggest.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/dev_perception/feedforward_neural_network.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx-PS2ktNV0pDH-ig1:-KqAM4V0fD4xTQTTeGLM:bf1fa2j) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L23)):*
<details><summary><i>Previously, nikos-tri wrote…</i></summary><blockquote>

Not sure if this was the spot where this comment belonged, since I don't see any code that looks similar. Maybe you were looking at an old version of the encoding function, which was adding 2. But, that was not due to the layers, it was due to needing two slots to store the number of rows and columns. But, the new version doesn't do that anymore due to annoyances with casting between AutoDiffXd and int.
</blockquote></details>

Based on a f2f discussion, I was referring to the line
```C++
DRAKE_THROW_UNLESS(W.size() == layers.size());
```
in `FeedforwardNeuralNetwork`. 

As we discussed, we decide to include the output layer, but not the input layer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 9 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 178 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqM8fvJQYyPXgAn6JrF:-KqM8fvJQYyPXgAn6JrG:bz3rq6n) ([raw file](https://github.com/robotlocomotion/drake/blob/71969d953924d5ac594353911477958f99a1e4e9/drake/dev_perception/feedforward_neural_network.cc#L178)):*
> ```cpp
>                                                       bias_indices_[index]);
> 
>   VectorX<T>* biasVector = new VectorX<T>(encodedVector.get_value());
> ```

`biasVector` -> `bias_vector`.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 186 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqM8nEfjLh2ABBVVRUM:-KqM8nEfjLh2ABBVVRUN:b1ulnor) ([raw file](https://github.com/robotlocomotion/drake/blob/71969d953924d5ac594353911477958f99a1e4e9/drake/dev_perception/feedforward_neural_network.cc#L186)):*
> ```cpp
> std::unique_ptr<BasicVector<T>> FeedforwardNeuralNetwork<
>     T>::EncodeWeightsToBasicVector(const MatrixX<T>& weights) const {
>   VectorX<T> dataVector(weights.size());
> ```

`data_vector`.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 188 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqM8p25HG0peXtTrX89:-KqM8p25HG0peXtTrX8A:b-7ropln) ([raw file](https://github.com/robotlocomotion/drake/blob/71969d953924d5ac594353911477958f99a1e4e9/drake/dev_perception/feedforward_neural_network.cc#L188)):*
> ```cpp
>   VectorX<T> dataVector(weights.size());
> 
>   int vectorIndex = 0;
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

nikos-tri commented:
Review status: 1 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 178 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqM8fvJQYyPXgAn6JrF:-KqQ-8R_OT-VVJ_VvYE3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71969d953924d5ac594353911477958f99a1e4e9/drake/dev_perception/feedforward_neural_network.cc#L178)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`biasVector` -> `bias_vector`.
</blockquote></details>

Done.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 186 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqM8nEfjLh2ABBVVRUM:-KqQ-7vqmt4g11SxaUNI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71969d953924d5ac594353911477958f99a1e4e9/drake/dev_perception/feedforward_neural_network.cc#L186)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`data_vector`.
</blockquote></details>

Done.

---

*[drake/dev_perception/feedforward_neural_network.cc, line 188 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqM8p25HG0peXtTrX89:-KqQ-7hzdpO6p_upc0bn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71969d953924d5ac594353911477958f99a1e4e9/drake/dev_perception/feedforward_neural_network.cc#L188)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

nikos-tri commented:
Review status: 1 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/dev_perception/feedforward_neural_network.h, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kpx-PS2ktNV0pDH-ig1:-KqQ-_5g8AcT3rgFB_Vg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/2871f4d6b37b3f30de813f2efa4895e8d49e939c/drake/dev_perception/feedforward_neural_network.h#L23)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Based on a f2f discussion, I was referring to the line
```C++
DRAKE_THROW_UNLESS(W.size() == layers.size());
```
in `FeedforwardNeuralNetwork`. 

As we discussed, we decide to include the output layer, but not the input layer.

</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 1 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/test/feedforward_neural_network_test.cc, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqQ0p2D_0Gm43Q6XQBc:-KqQ0p2D_0Gm43Q6XQBd:bgl2vx7) ([raw file](https://github.com/robotlocomotion/drake/blob/10766525ae66faac189f8a63a9057834d6273006/drake/dev_perception/test/feedforward_neural_network_test.cc#L74)):*
> ```cpp
> // Test that the NN is correctly loading and extracting its parameters from
> // the Context
> GTEST_TEST(FeedforwardNeuralNetworkTest, ParameterStorage) {
> ```

BTW, the lines
```C++
MatrixXd W1 = MatrixXd::Random(3, 3);
MatrixXd W2 = MatrixXd::Random(7, 3);
MatrixXd W3 = MatrixXd::Random(2, 7);
std::vector<MatrixXd> W;
  W.push_back(W1);
  W.push_back(W2);
  W.push_back(W3);
 
  VectorXd B1 = VectorXd::Random(3);
  VectorXd B2 = VectorXd::Random(7);
  VectorXd B3 = VectorXd::Random(2);
  std::vector<VectorXd> B;
  B.push_back(B1);
  B.push_back(B2);
  B.push_back(B3);
 
  FeedforwardNeuralNetwork<double> dut(W, B, generateFullyConnectedLayers(3),
                                       generateReluNonlinearities(3));
 
  unique_ptr<Context<double>> context = dut.CreateDefaultContext();
```
are duplicated in all three `GTEST_TEST`, you can consider to create a test
```C++
class FeedforwardNeuralNetworkTest : public ::testing::Test {
  public:
    FeedforwardNeuralNetworkTest
    : W1_{MatrixXd::Random(3, 3)},
      W2_{MatrixXd::Random(7, 3)},
      W3_{MatrixXd::Random(2, y)},
      W_{{W1, W2, W3}},
      B1_{VectorXd::Random(3)},
      B2_{VectorXd::Random(7)},
      B3_{VectorXd::Random(2)},
      B_{{B1_, B2_, B3_}},
      dut_{W_, B_, generateFullyConnectedLayers(3), generateReluNOnlinearities(3)},
      context_{dut_.CreateDefaultContext()} {
    }

  protected:
    MatrixXd W1_;
    MatrixXd W2_;
    MatrixXd W3_;
    std::vector<MatrixXd> W_;
    VectorXd B1_;
    VectorXd B2_;
    VectorXd B3_;
    std::vector<VectorXd> B_;
    FeedforwardNeuralNetwork<double> dut_;
    unique_ptr<Context<double>> context_;
}
```
And your test can be written as
```C++
TEST_F(FeedforwardNeuralNetworkTest, ParamStorage) {
  EXPECT_EQ(W1_, *(dut_.get_weight_matrix(0, *context)));
  EXPECT_EQ(W2_, *(dut_.get_weight_matrix(1, *context)));
  EXPECT_EQ(W3_, *(dut_.get_weight_matrix(2, *context)));
 
  // cout << "Expected: " << endl << B1 << endl;
  // cout << "Recovered: " << endl << *(dut.get_bias_vector(0, *context));
  EXPECT_EQ(B1_, *(dut_.get_bias_vector(0, *context)));
  EXPECT_EQ(B2_, *(dut_.get_bias_vector(1, *context)));
  EXPECT_EQ(B3, *(dut.get_bias_vector(2, *context)));
}
```
same for the other `GTEST_TEST` (ShapeParameters and ToAutodiffTest).

Another solution is to put `ParamStorage`, `ShapeParameters` and `ToAutodiffTest` into one `GTEST_TEST`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 6 files at r1.
Review status: 1 of 9 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/dev_perception/test/feedforward_neural_network_test.cc, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqQ0p2D_0Gm43Q6XQBc:-KqTplhDhZJy61RTffQT:bbzoykz) ([raw file](https://github.com/robotlocomotion/drake/blob/10766525ae66faac189f8a63a9057834d6273006/drake/dev_perception/test/feedforward_neural_network_test.cc#L74)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

BTW, the lines
```C++
MatrixXd W1 = MatrixXd::Random(3, 3);
MatrixXd W2 = MatrixXd::Random(7, 3);
MatrixXd W3 = MatrixXd::Random(2, 7);
std::vector<MatrixXd> W;
  W.push_back(W1);
  W.push_back(W2);
  W.push_back(W3);
 
  VectorXd B1 = VectorXd::Random(3);
  VectorXd B2 = VectorXd::Random(7);
  VectorXd B3 = VectorXd::Random(2);
  std::vector<VectorXd> B;
  B.push_back(B1);
  B.push_back(B2);
  B.push_back(B3);
 
  FeedforwardNeuralNetwork<double> dut(W, B, generateFullyConnectedLayers(3),
                                       generateReluNonlinearities(3));
 
  unique_ptr<Context<double>> context = dut.CreateDefaultContext();
```
are duplicated in all three `GTEST_TEST`, you can consider to create a test
```C++
class FeedforwardNeuralNetworkTest : public ::testing::Test {
  public:
    FeedforwardNeuralNetworkTest
    : W1_{MatrixXd::Random(3, 3)},
      W2_{MatrixXd::Random(7, 3)},
      W3_{MatrixXd::Random(2, y)},
      W_{{W1, W2, W3}},
      B1_{VectorXd::Random(3)},
      B2_{VectorXd::Random(7)},
      B3_{VectorXd::Random(2)},
      B_{{B1_, B2_, B3_}},
      dut_{W_, B_, generateFullyConnectedLayers(3), generateReluNOnlinearities(3)},
      context_{dut_.CreateDefaultContext()} {
    }

  protected:
    MatrixXd W1_;
    MatrixXd W2_;
    MatrixXd W3_;
    std::vector<MatrixXd> W_;
    VectorXd B1_;
    VectorXd B2_;
    VectorXd B3_;
    std::vector<VectorXd> B_;
    FeedforwardNeuralNetwork<double> dut_;
    unique_ptr<Context<double>> context_;
}
```
And your test can be written as
```C++
TEST_F(FeedforwardNeuralNetworkTest, ParamStorage) {
  EXPECT_EQ(W1_, *(dut_.get_weight_matrix(0, *context)));
  EXPECT_EQ(W2_, *(dut_.get_weight_matrix(1, *context)));
  EXPECT_EQ(W3_, *(dut_.get_weight_matrix(2, *context)));
 
  // cout << "Expected: " << endl << B1 << endl;
  // cout << "Recovered: " << endl << *(dut.get_bias_vector(0, *context));
  EXPECT_EQ(B1_, *(dut_.get_bias_vector(0, *context)));
  EXPECT_EQ(B2_, *(dut_.get_bias_vector(1, *context)));
  EXPECT_EQ(B3, *(dut.get_bias_vector(2, *context)));
}
```
same for the other `GTEST_TEST` (ShapeParameters and ToAutodiffTest).

Another solution is to put `ParamStorage`, `ShapeParameters` and `ToAutodiffTest` into one `GTEST_TEST`.

</blockquote></details>

FYI, the use of `Eigen::Random` is not allowed by http://drake.mit.edu/code_review_checklist.html#is-your-code-deterministic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 10 files at r2, 6 of 12 files at r5.
Review status: 4 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/perception/dev/feedforward_neural_network.cc, line 245 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqcsItEdvXnAsalfEhp:-KqcsItEdvXnAsalfEhq:b-l1mtk1) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/feedforward_neural_network.cc#L245)):*
> ```cpp
> }
> 
> template class FeedforwardNeuralNetwork<double>;
> ```

Also need to explicitly instantiate the case for `AutodiffScalar`. The CI failure complains about this. https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=728768

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 247 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqcsrRQyTO4Se3mu-2H:-KqcsrRQyTO4Se3mu-2I:bdtcbfg) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L247)):*
> ```cpp
> 
> // Helper function to generate a vector of Connected layers
> std::vector<LayerType> generateFullyConnectedLayers(int numLayers) {
> ```

Based on the code standard, we should use `GenerateFullyConnectedLayers`. The first letter should be a capital letter.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 257 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kqcszii0FgPkR2IViKA:-Kqcszii0FgPkR2IViKB:b-rqxvo7) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L257)):*
> ```cpp
> 
> // Helper function to generate a vector of Relu nonlinearities
> std::vector<NonlinearityType> generateReluNonlinearities(int numLayers) {
> ```

ditto.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 267 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kqct2q9NNMHIIkiOCb3:-Kqct2q9NNMHIIkiOCb4:b-if64pr) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L267)):*
> ```cpp
> 
> // Helper function to compare a MatrixX<double> and a MatrixX<AutoDiffXd>
> bool CompareMats(MatrixX<double> Mdouble, MatrixX<AutoDiffXd> Mautodiff) {
> ```

You can use pass by const reference, instead of pass by value here. Namely `MatrixX<double>` -> `const MatrixX<double>&`

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 277 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqctG4tIYvtEk7DVeZX:-KqctG4tIYvtEk7DVeZY:b-dwskh7) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L277)):*
> ```cpp
>       double thisAdValue = (Mautodiff(i, j)).value();
> 
>       if (thisAdValue != Mdouble(i, j)) {
> ```

Comparing two double values directly might not be what you want. You may consider to specify a threshold here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

nikos-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/perception/dev/feedforward_neural_network.cc, line 245 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqcsItEdvXnAsalfEhp:-Kqd1GtbVsX7EoTDIdJJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/feedforward_neural_network.cc#L245)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Also need to explicitly instantiate the case for `AutodiffScalar`. The CI failure complains about this. https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=728768
</blockquote></details>

Done.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqQ0p2D_0Gm43Q6XQBc:-Kqd22SkZM9VojtG-IE5:b-v74072) ([raw file](https://github.com/robotlocomotion/drake/blob/10766525ae66faac189f8a63a9057834d6273006/drake/perception/dev/test/feedforward_neural_network_test.cc#L74)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, the use of `Eigen::Random` is not allowed by http://drake.mit.edu/code_review_checklist.html#is-your-code-deterministic.
</blockquote></details>

Removed use of MatrixXd::Random. I'll pass on the test class for now, since it seems to me that the reusability of that code would be limited. Not even reusable for NN tests with different numbers of layers.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 247 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqcsrRQyTO4Se3mu-2H:-Kqcuv6Hm9QSZ2ln85Mn:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L247)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Based on the code standard, we should use `GenerateFullyConnectedLayers`. The first letter should be a capital letter.
</blockquote></details>

Done

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 257 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kqcszii0FgPkR2IViKA:-Kqcun88SQpWL9QX_kR4:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L257)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

ditto.
</blockquote></details>

Done

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 267 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kqct2q9NNMHIIkiOCb3:-KqcuoJUQUVtK4pyhXDp:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L267)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

You can use pass by const reference, instead of pass by value here. Namely `MatrixX<double>` -> `const MatrixX<double>&`
</blockquote></details>

Done

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 277 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqctG4tIYvtEk7DVeZX:-Kqcu_QgvnU0IOUKRCN_:b-f7hsyz) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L277)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Comparing two double values directly might not be what you want. You may consider to specify a threshold here.
</blockquote></details>

It should be the exact same double value because there is no arithmetic, just copying the number around. So, there is no problem with precision. However, added an optional function parameter so that it can be re-used  in the future when there is arithmetic involved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with minor comments. 

---

Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 247 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqcsrRQyTO4Se3mu-2H:-Kqd5VOwb9gBtWuIYNFf:bt8vxd3) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L247)):*
<details><summary><i>Previously, nikos-tri wrote…</i></summary><blockquote>

Done
</blockquote></details>

`numLayers` -> `num_layers`.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 257 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kqcszii0FgPkR2IViKA:-Kqd5ZYJoQUC6uJJUOEG:b-7ropln) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L257)):*
<details><summary><i>Previously, nikos-tri wrote…</i></summary><blockquote>

Done
</blockquote></details>

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-Kqd5dwPhNaLhTNKKhqU:bjk42yk)*
<!-- Sent from Reviewable.io -->

nikos-tri commented:
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 247 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-KqcsrRQyTO4Se3mu-2H:-Kqd6Q2kKIVZem3_hYiV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L247)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

`numLayers` -> `num_layers`.
</blockquote></details>

Done.

---

*[drake/perception/dev/test/feedforward_neural_network_test.cc, line 257 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-Kqcszii0FgPkR2IViKA:-Kqd6QjlgURXoGI4z6NK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/a3fb13298d5f23bc15819ea380a01276a12f77d4/drake/perception/dev/test/feedforward_neural_network_test.cc#L257)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
+(status: curate commits before merging) 

---

Review status: 3 of 5 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6621#-:-KqdECO0ar_E6Bvd9N_N:b-s794vx)*
<!-- Sent from Reviewable.io -->
